### PR TITLE
perf(over window): locality-based range cache for general over window executor

### LINKED ISSUE #11502: Better cache strategy for general `OverWindowExecutor`

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR implements a _range cache_ for general `OverWindowExecutor`. Instead of retaining everything in an accessed over partition in memory, the range cache only caches recently accessed range of rows in each accessed partition. This can significantly reduce memory footprint for window function calls with bounded window frames like `lead(x) over (...)`, `sum(x) over (partition by ... order by ... rows 10 preceding)`.

This fixes #11502.

A session variable called `RW_STREAMING_OVER_WINDOW_CACHE_POLICY` is introduced to control the cache policy:

- `full`: default policy, cache all entries in each accessed partition, just like before
- `recent`: cache only recently accessed range
- `recent_first_n`: cache only the first several entries in recently accessed range
- `recent_last_n`: cache only the last several entries in recently accessed range

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- ~~[ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).~~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [x] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- ~~[ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))~~

## Documentation

- ~~[ ] My PR contains user-facing changes.~~

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/stream/src/executor/over_window/over_partition.rs [hunk 1]
```diff
@@ -0,0 +1,1256 @@
...
+        return vec![];
+    }
+
+    vec![(
+        first_frame_start,
+        first_curr_key,
+        last_curr_key,
+        last_frame_end,
+    )]
+}
```
Comment (stdrc) [inline src/stream/src/executor/over_window/over_partition.rs:866]: These LoC are basically moved from `general.rs` without change.

File: src/stream/src/executor/over_window/over_partition.rs [hunk 2]
```diff
@@ -0,0 +1,1256 @@
...
+        let delta = create_delta!((2, Delete), (3, Insert), (4, Delete));
+        let calls = vec![create_call(Frame::rows(
+            FrameBound::CurrentRow,
+            FrameBound::CurrentRow,
+        ))];
+        assert_ranges_eq(
+            find_affected_ranges(&calls, DeltaBTreeMap::new(&cache, &delta)),
+            [(3.into(), 3.into(), 3.into(), 3.into())],
+        );
+    }
```
Comment (stdrc) [inline src/stream/src/executor/over_window/over_partition.rs:1132]: These are not related to the range cache.

File: src/stream/src/executor/over_window/over_partition.rs [hunk 3]
```diff
@@ -0,0 +1,1256 @@
...
+        table: &'_ StateTable<S>,
+        delta: &'cache PartitionDelta,
+    ) -> StreamExecutorResult<(
+        DeltaBTreeMap<'cache, CacheKey, OwnedRow>,
+        Vec<(
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+            &'cache CacheKey,
+        )>,
```
Comment (stdrc) [inline src/stream/src/executor/over_window/over_partition.rs:257]: Return keys with lifetime `'cache` which is corresponding to parameter `delta` and `self` (`'s: 'cache`) to give the caller confidence that the key is guaranteed to exist in the `DeltaBTreeMap`.

File: src/stream/src/executor/over_window/over_partition.rs [hunk 4]
```diff
@@ -0,0 +1,1256 @@
...
+                let left_most = self.cache_real_first_key().unwrap_or(delta_first).clone();
+                self.extend_cache_leftward_by_n(table, &left_most).await?;
+            }
+            if right_reached_sentinel {
+                tracing::debug!(partition=?self.this_partition_key, "partition cache right extension triggered");
+                let right_most = self.cache_real_last_key().unwrap_or(delta_last).clone();
+                self.extend_cache_rightward_by_n(table, &right_most).await?;
+            }
+            tracing::debug!(partition=?self.this_partition_key, "partition cache extended");
+        }
```
Comment (stdrc) [inline src/stream/src/executor/over_window/over_partition.rs:312]: Terminateability: `extend_cache_leftward_by_n` and `extend_cache_rightward_by_n` keep pushing the cache to the boundary of current partition. In these two methods, when any side of boundary is reached, the sentinel key will be removed, so finally `self::find_affected_ranges` will return ranges without any sentinels.

File: src/stream/src/executor/over_window/over_partition.rs [hunk 5]
```diff
@@ -0,0 +1,1402 @@
...
+                if !left_reached_sentinel && !right_reached_sentinel {
+                    // all affected ranges are already cached, we're done
+                    return Ok((DeltaBTreeMap::new(cache_inner, delta), ranges));
+                }
+
+                (left_reached_sentinel, right_reached_sentinel)
+            };
+
+            if left_reached_sentinel {
+                // TODO(rc): should count cache miss for this, and also the below
```
Comment (stdrc) [inline src/stream/src/executor/over_window/over_partition.rs:447]: Will do cache miss counting in future PR, to avoid making this PR too large.

File: src/common/src/session_config/mod.rs [hunk 6]
```diff
@@ -473,6 +476,13 @@ pub struct ConfigMap {
     standard_conforming_strings: StandardConformingStrings,
 
     streaming_rate_limit: StreamingRateLimit,
+
+    /// Cache policy for partition cache in streaming over window.
+    /// Can be "full", "recent", "recent-first-n" or "recent-last-n".
+    #[educe(Default(
+        expression = "ConfigString::<STREAMING_OVER_WINDOW_CACHE_POLICY>(String::from(\"full\"))"
+    ))]
+    streaming_over_window_cache_policy: StreamingOverWindowCachePolicy,
```
Comment (st1page) [inline src/common/src/session_config/mod.rs:485]: maybe change it to an enum, I think you can refer to `QueryMode`

**Review by st1page — COMMENTED** [review]
rest generally LGTM and we can maintain the CachePolicy the same with previous

**Review by st1page — APPROVED** [review]

**Comment by stdrc** [comment]
The workflow requires members of `Cargo.lock` team to approve because there's one change in `Cargo.lock` file. Plz help take a look @BugenZhao @TennyZhuang @wangrunji0408

**Review by wangrunji0408 — APPROVED** [review]
